Interview judgement branch

daido1976 opened the pull request:
#やりたかったこと
---
- 自分以外のユーザーの一覧をトップページに表示する
- 面接官用の面接日程一覧ページを作る
- 面談日程を承認or拒否できるようにする
- 過去の面接日程は承認できないようにする
- 同じユーザーは同じ面接日程を登録できないようにする

#やったこと
----
- トップページに自分以外のユーザー(where.not(id:current_user.id))の一覧を追加
- 面接官用の面接日程一覧をpartialで出し分ける
- 面接日程の承認機能(ブラウザからhidden_fieldでvalue:"approval"を渡し、controllerのupdateアクション内でそれ以外の日程を"rejection"とする)を実装
- 過去の面接日程は承認できないよう、面接官用のviewにif else文を追加
- 同じユーザーは同じ面接日程を登録できないよう、modelにvalidationを設定

#動作確認方法
--- 
- [0] 事前にいくつか面接日程を登録しておく(ユーザー1とする)
- [1] 別ユーザーを新規で作成(sign-up)する(ユーザー2とする)
- [2] トップページより事前に面接日程を登録したユーザーの面接一覧ページ(Interviews)にいく(ユーザー2)
- [3] 任意の面接日程の保存(save)ボタンを押す(ユーザー2)
- [4] judgementが「承認」になっているのを確認(ユーザー2)
- [5] ユーザー1でサインインし直し、自分の面接一覧画面でも任意の面接日程が「承認」になっていることを確認(ユーザー1)

#その他
---
面接日程承認機能の「承認日以外を全て却下として保存する」処理はもっとスマートに書ける気がしています。
「もっとこう書いた方がいいよ！」みたいなところがあればドンドンご指摘ください！
お手数おかけしますが、よろしくお願い致します！

Commit:
Add users-list to top-page

Commit:
Interviews judgement

Commit:
rake assets:precompile

tmd45 commented on app/controllers/interviews_controller.rb:
```diff
@@ -27,6 +26,7 @@ def edit
   def update
     @interview = @user.interviews.find(params[:id])
     if @interview.update(interview_params)
+      @user.interviews.where.not(id: @interview.id).update_all(judgement: "rejection") unless @user == current_user
```
`@user.interviews.where.not(id: @interview.id).update_all(judgement: "rejection")` が更新に失敗したら、interview の更新結果はどのようになるでしょうか？

いくつかの更新処理がかならず同時に行われること、
いずれかが失敗したら全て失敗することなどを制御するには `transaction` の仕組みを使うと良いです。

また、controller の中で処理が複雑になってきたら、model にその処理をするメソッドを移して、controller からそのメソッドを呼ぶように書くのもひとつの方法です。

tmd45 replied:
この `update` メソッドは、現在２つの仕事をしています。

- `current_user` の場合
    - `@user`（＝ `current_user`）の `@interview` の `:interview_datetime` を更新する（judgement を更新することはない）
- `current_user` 以外の場合
    - `@user` の `@interview` の `:judgement` を更新する（interview_datetime を更新することはない）
    - `@user.interviews`（`@interview` 以外）の `:judgement` を更新する
        - この２つの処理は同時に行いたい

一度書いたコードや動作を眺めてみて整理したら、それに合わせてコードをリファクタリングしてみましょう 💪 （上の整理の仕方は一例です）

tmd45 replied:
もっと言うと、１つのメソッド（#update）でこのように２つの仕事をしはじめたら、メソッドを分けたほうが良いかもしれないサインです。
ここまで仕事がはっきり分かれているのであれば、action を別にしてしまうのも手です 😎

tmd45 reviewed (commented):
> 面接日程承認機能の「承認日以外を全て却下として保存する」処理はもっとスマートに書ける気がしています

一度動くように書いたコードを、遠目から眺めて整理してみるとよいかもしれません。
というようなことを以下にコメントしましたのでご確認下さい！

tmd45 commented:
> 面接日程承認機能の「承認日以外を全て却下として保存する」処理はもっとスマートに書ける気が

https://github.com/daido1976/e-navigator/pull/3#discussion_r170522392 で `transaction` の話を書きましたが、
ActiveRecord の `after_update` というコールバック処理を用いることができるかもしれません 💡

Commit:
add judgement action

current_user 以外の場合に処理をするJudgement actionのルーティング、メソッド定義を行いました。
レビューいただいた通り、judgement actionではinterview_datetimeを更新することはないので、明示的にstrong
parameters(judgement_params)も分けてみましたが、明示する必要なければ元に戻します。

Commit:
after_update :judgement_others

Judgementの処理をmodel内にafter_updateとして分割し、transactionとして扱われるようにしました。

daido1976 commented:
@tmd45 
レビューいただいた部分を修正しましたので再度ご確認お願い致します！
(コード的にはあまり短くなっておらず、、レビューいただいた趣旨を汲み取れていなかったら申し訳ございません、、)

tmd45 commented:
ばっちり汲み取っていただけてます！ 😄 👍 

> コード的にはあまり短くなっておらず

不要な機能は実装しないようにする、無駄なコードは書かないようにする
という意味でコードを減らすことはありますが
大事なのは可読性、保守性です！

条件分岐を読み違えないようにワンライナーでの記述をやめたり、
条件式だけを別メソッドに切り出したり、
そうすると実際のコード量・行数は少しですが増えます。
ですが大事なのは可読性、保守性です！（大事なことなのでｒｙ 😁 

無駄なコードを減らすのは、冗長でも全体を一度書ききってからのほうがよいです。
全体を見渡して、ここは１つにできそうとか、別の書き方にすればスッキリするとか、そういうリファクタリングを行うのが良い流れだと思います ✨ 
最初からリファクタリングをやってしまうのは、結果的にリファクタリングにならないかもしれません 👻

daido1976 commented:
@tmd45 
大事なことを2回も言っていただいてありがとうございます！笑
可読性、保守性意識していきますね！
ラスト頑張りますので、引き続きよろしくお願い致します！
